Move CardsList to typescript

max-voronov opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor

## Description

Hello everyone. If it's possible, I would like to help with moving to typescript.
In this PR the `CardsList` component was moved to typescript and then converted into a functional component.

## Related Tickets & Documents

#5027 
 
## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

There aren't any UI changes

kravets-levko commented:
Hi @max-voronov! Thank you for your contribution! I'd like to ask you about one more change to this PR: could you try to add a commit that just renames `.jsx` file to `.tsx`, and then rebase everything else on top of it? It's totally not necessary, but this will allow GIT to track a rename, and (I hope) it will display a proper diff here instead of `removed` + `created` changes. Thanks!

max-voronov commented:
> Hi @max-voronov! Thank you for your contribution! I'd like to ask you about one more change to this PR: could you try to add a commit that just renames `.jsx` file to `.tsx`, and then rebase everything else on top of it? It's totally not necessary, but this will allow GIT to track a rename, and (I hope) it will display a proper diff here instead of `removed` + `created` changes. Thanks!

Hi @kravets-levko and thanks for replying!

I did what you asked me to, but this won't change the total diff you see under the `Files changed` tab.
It will still be showed as `removed` + `created` changes, so I think this commit is redundant and better to be removed.

Please let me know if I should remove it or not.

kravets-levko commented:
I hoped that it will work :slightly_smiling_face: Anyway, thank you, and sorry that I bothered you :bow:

gabrieldutra commented:
No need to revert @max-voronov, in the end we will squash the commit anyway 😅

max-voronov commented:
> I hoped that it will work 🙂 Anyway, thank you, and sorry that I bothered you 🙇

You didn't bother me 😃
I hoped it will help too but I guess there are too many changes to the same file that include indentation changes, so GIT sees it as `removed` + `created` change even tough there are some lines that their only change is indentation.

gabrieldutra commented on client/app/components/cards-list/CardsList.tsx:
```diff
@@ -0,0 +1,73 @@
+import { includes, isEmpty } from "lodash";
+import PropTypes from "prop-types";
+import React, { useState } from "react";
+import Search from "antd/lib/input";
```
This will make it use the `Input` component instead of the `Input.Search`. Let's rename it to `Input` here and use `Input.Search` as the component in https://github.com/getredash/redash/blob/d6025f9186d8db8e45b79a6fc4b0afd0855c19fb/client/app/components/cards-list/CardsList.tsx#L42

I think the search is actually not used in the app, but let's keep it as it was in the Class component.

max-voronov replied:
Fixed in 83b13dbdb58d43f654223f1da198e387e1bc9ada

gabrieldutra reviewed (commented):
Overall looks good, thanks @max-voronov 😁. I left just one small comment.

The component seems better and simplified as a Functional component. But I'm curious, did you have any specific reason to move it to TS?

I guess we haven't actually decided how to proceed with TS in the codebase at this point, we added #5027 so that we have support for it (there were a few spots that we needed it), but I'm not sure I'd go around refactoring existing components (unless there's a reason for so). For me there's no problem at all with this component for example, but some of them can be quite complicated 😅.

max-voronov commented:
> Overall looks good, thanks @max-voronov 😁. I left just one small comment.
> 
> The component seems better and simplified as a Functional component. But I'm curious, did you have any specific reason to move it to TS?
> 
> I guess we haven't actually decided how to proceed with TS in the codebase at this point, we added #5027 so that we have support for it (there were a few spots that we needed it), but I'm not sure I'd go around refactoring existing components (unless there's a reason for so). For me there's no problem at all with this component for example, but some of them can be quite complicated 😅.

Thanks for the review 😃 

I assumed that because you added Typescript support in #5027, you would like to convert to Typescript.
There is also [this discussion](https://discuss.redash.io/t/switch-to-typescript/3075) with some good points in favour of using Typescript.

If you decide that you would like to convert the whole project to Typescript then I'd like to help with the process 😃

gabrieldutra commented:
Yeah, I guess it's time to revisit that discussion now that we actually have TS support and some of the code in TS too 😁.

@max-voronov thank you very much for your interest and for your work here 🎉.

kravets-levko commented on client/app/components/cards-list/CardsList.tsx:
```diff
@@ -0,0 +1,73 @@
+  showSearch?: boolean;
+}
+
+function ListItem(item: CardsListItem, keySuffix: string) {
+  return (
+    <Link key={`card${keySuffix}`} className="visual-card" onClick={item.onClick} href={item.href}>
+      <img alt={item.title} src={item.imgSrc} />
+      <h3>{item.title}</h3>
+    </Link>
+  );
+}
```
Overall code looks good, just a little last thing: let's convert this to a component, because it already looks like a component, just need to change arguments to props, and update usage

max-voronov replied:
Fixed in c39f254b2bef2f5962039ef8e1e60fa697d0a53e

Thanks for the review 😃

kravets-levko reviewed (changes requested):
Please see my comment - that's the last little change I'd like to see here before I'll merge this. Thanks!

kravets-levko commented:
Thank you @max-voronov!
